### PR TITLE
eq-1190 allow conditional date piping

### PR DESCRIPTION
**Pull request by berianjones**
this change allows a choice between submitted dates or default to metadata dates

### What is the context of this PR?
New def format_cond_date added to jinja_filters.py and change reflected in template_renderer.py
This allows the use of default dates from metadata

### How to review 
I modified an existing json (not checked in), made the date entry fields non-mandatory, created an alias for the period-from and period-to answers (cf with census-household.json). I then added the following to a title held in the subsequent questions in the json 
```
{{[answers.period_to, exercise.start_date]|format_cond_date}}
{{[answers.period_from, exercise.end_date]|format_cond_date}}
```
where period_from and period_to are the aliases and exercise.start_date/end_date are the metadata values. Tested for blank dates, invalid dates

### CODE REVIEW

File: app/jinja_filters.py
Comment (jonnyshaw89) [inline app/jinja_filters.py]: I can't see any reason to abbreviate this name?

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]
I think you also need to add some tests for this new filter.

**Comment by jonnyshaw89** [comment]
Did you accidentally commit `yarn.lock`?

File: app/jinja_filters.py
Comment (jonnyshaw89) [inline app/jinja_filters.py]: the function name in the comment is now incorrect

**Comment by dwyeradam** [comment]
I agree it would be worth having a test for this. There is already a test_dates.json that could be expanded to include an extra page with piping.

**Comment by collisdigital** [comment]
Please use commit messages to describe the change being made, it's making it hard to see what's going on 
image

File: app/jinja_filters.py
Comment (ajmaddaford) [inline app/jinja_filters.py]: You should be able to replace these lines with calls to `format_str_as_short_date` e.g. `format_str_as_short_date(dates[0])`

**Review by ajmaddaford — CHANGES_REQUESTED** [review]

File: app/jinja_filters.py
Comment (jonnyshaw89) [inline app/jinja_filters.py]: why do the 2 return types use different methods to format the date?

File: package.json
Comment (jonnyshaw89) [inline package.json]: What is this used for?

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: app/jinja_filters.py
Comment (berianjones) [reply app/jinja_filters.py]: I think this is due to the date[1] being derived from the metadata. It errors if I use the same function for both

File: package.json
Comment (berianjones) [reply package.json]: Nothing, it must have been a consequence f the yarn.lock accidental commit

**Comment by berianjones** [comment]
Added tests and addressed comments, reverted changes to yarn.lock and packages.json

File: app/jinja_filters.py
Comment (jonnyshaw89) [reply app/jinja_filters.py]: I think these should both use the same function otherwise the returned date will be formatted differently.
If the 2 values are in different formats then that just needs to be handled. 
This function shouldn't be implemented to just handle some data coming from metadata.

**Comment by jonnyshaw89** [comment]
@berianjones the `yarn.lock` and `package.json` changes are still in this PR?

File: app/jinja_filters.py
Comment (ajmaddaford) [reply app/jinja_filters.py]: Dates in answers are stored as a string in a `dd/mm/yyyy` format, whereas dates in the metadata and survey contexts are stored as Python date objects. 

As this method should be able to compare any dates, it should deal with both date formats in either argument. I think for each date it would be fine to check if it's a Python date object and if not assume it's a date in a `dd/mm/yyyy` format.

This has highlighted some inconsistencies in the way that we handle dates that we should review.

File: tests/app/test_jinja_filters.py
Comment (ajmaddaford) [inline tests/app/test_jinja_filters.py]: If we adapt the method to work with dates in either format for both arguments to `format_conditional_date`, then we'll need additional tests to deal with all of the combinations.

**Review by ajmaddaford — CHANGES_REQUESTED** [review]

**Comment by collisdigital** [comment]
I'm still seeing a commits on this branch with the same commit message, please be descriptive when committing changes as it makes reviewing the history hard otherwise

**Comment by berianjones** [comment]
Ready for review

File: app/jinja_filters.py
Comment (ajmaddaford) [reply app/jinja_filters.py]: When documenting the arguments and return values in the docstring we should follow the standard used elsewhere e.g. https://github.com/ONSdigital/eq-survey-runner/blob/master/app/data_model/answer_store.py#L153

File: app/jinja_filters.py
Comment (ajmaddaford) [inline app/jinja_filters.py]: As discussed, we should assume this method is called with only two date 'formats' - a `dd/mm/yyyy` string date from the answer store, or a Python Date object from the metadata. Then, for each date argument we can just test if it's already a Date object and if not convert it to one (assuming a string in the format of `dd/mm/yyyy`). 
The Date object can then be output as a date string in the appropriate format using the `format_date` method.

File: tests/app/test_jinja_filters.py
Comment (ajmaddaford) [inline tests/app/test_jinja_filters.py]: We should have more unit tests. Currently there are no tests for when dates passed to this method are Python Date objects, and there are no tests with just the first date argument set.

File: tests/app/test_jinja_filters.py
Comment (ajmaddaford) [inline tests/app/test_jinja_filters.py]: The method name should reflect what the method is testing, so something like `test_format_conditional_date_first_arg_none` would be better. 

It might also be worth considering creating a data structure to represent the test variations and loop over it e.g. 
```
test_dates = [
  ['03/11/2017', '04/08/2018', '11 March 2017'],
  [None, '10/12/17', '12 October 2017']
]
```
or possibly clearer using dicts:
```
test_dates = [
  {
    'date1': '03/11/2017', 
    'date2': '04/08/2018', 
    'result': '11 March 2017'
  },
  {
    'date1': None, 
    'date2': '10/12/17', 
    'result': '12 October 2017'
  }
]
```

**Review by ajmaddaford — CHANGES_REQUESTED** [review]

**Comment by berianjones** [comment]
Ready for review. New tests added and code changes made. All tests passed

File: app/jinja_filters.py
Comment (jonnyshaw89) [reply app/jinja_filters.py]: @berianjones this comment is still incorrect

File: app/jinja_filters.py
Comment (jonnyshaw89) [inline app/jinja_filters.py]: I wonder if this logic could go inside the `format_date` function?

File: package.json
Comment (jonnyshaw89) [reply package.json]: @berianjones this new dependency is still here. if we don't use it, it should be removed.

File: yarn.lock
Comment (jonnyshaw89) [inline yarn.lock]: @berianjones given we shouldn't have any changes to the node dependencies then there is no reason that the lock file should be updated.

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: app/jinja_filters.py
Comment (ajmaddaford) [inline app/jinja_filters.py]: I think we should raise an Exception here if date is None, something like:
```
if date is None:
    raise Exception('No dates passed to format_conditional_date filter')
```

File: tests/app/test_jinja_filters.py
Comment (ajmaddaford) [inline tests/app/test_jinja_filters.py]: This is a little hard to read, I think it would be improved if:
- Each item was on it's own line
- The `datetime.strptime` method calls were replaced with `datetime.date` calls

Something like:
```
datelist = [
    ('12/01/2016', '12/02/2016', '12 January 2016'), 
    ('23/12/2017', None, '23 December 2017'),
    (None, datetime.date(2017, 10, 12), '12 October 2017'),
    (datetime.date(2019, 10, 12), datetime.date(2017, 9, 12), '12 October 2019'),
    (datetime.date(2018, 10, 12), datetime.date(2017, 9, 12), '12 October 2018'),
    (None, datetime.date(1000, 1, 1), '1 January 1000')
]
```
This also more clearly highlights some missing tests:
```
(None, '23/12/2017', '23 December 2017')
(datetime.date(2017, 10, 12), None, '12 October 2017'),
('23/12/2017', datetime.date(2017, 10, 12), '23 December 2017')
(datetime.date(2017, 10, 12), '23/12/2017', '12 October 2017')
```
And some duplicate tests:
```
(None, datetime.date(1000, 1, 1), '1 January 1000')
(datetime.date(2018, 10, 12), datetime.date(2017, 9, 12), '12 October 2018')
```

File: tests/app/test_jinja_filters.py
Comment (ajmaddaford) [inline tests/app/test_jinja_filters.py]: There's no need to create the dates variable, can't we just call `format_conditional_date([date, date2])`?

File: tests/app/test_jinja_filters.py
Comment (ajmaddaford) [inline tests/app/test_jinja_filters.py]: I think it would be just as clear to test `format_conditional_date([None, None])`

File: data/en/test_conditional_dates.json
Comment (ajmaddaford) [inline data/en/test_conditional_dates.json:1]: I think if we're adding a new survey to test the conditional dates it should be the minimal content required, this looks like it's a copy of the original test_dates survey with an additional set of dates and an extra page to test the conditional date filter. We should remove all date questions that don't relate to testing the conditional date filter.

File: data/en/test_dates.json
Comment (ajmaddaford) [inline data/en/test_dates.json]: This just looks like formatting changes, running `yarn format` should remove these unnecessary changes.

File: tests/functional/spec/dates.spec.js
Comment (ajmaddaford) [inline tests/functional/spec/dates.spec.js]: If a page has been added to the survey, then a page object should be generated and this submit should be called on that.

**Review by ajmaddaford — CHANGES_REQUESTED** [review]

File: app/jinja_filters.py
Comment (jonnyshaw89) [reply app/jinja_filters.py]: We should probably also validate for the existence of `dates[1]`

**Comment by berianjones** [comment]
Code and tests changed as requested in reviews

**Comment by berianjones** [comment]
All revisions made and new new_functional test added - ready for review

File: tests/new_functional/spec/dates_conditional.spec.js
Comment (ajmaddaford) [inline tests/new_functional/spec/dates_conditional.spec.js]: test_dates should be test_conditional_dates

File: tests/new_functional/pages/surveys/dates_conditional/summary.page.js
Comment (ajmaddaford) [inline tests/new_functional/pages/surveys/dates_conditional/summary.page.js]: Spacing is a bit off here

File: tests/functional/spec/dates.spec.js
Comment (ajmaddaford) [inline tests/functional/spec/dates.spec.js]: These extra new lines should be removed

File: tests/functional/pages/surveys/dates/dates-answers.page.js
Comment (ajmaddaford) [inline tests/functional/pages/surveys/dates/dates-answers.page.js]: I don't think there should be any changes to anything in tests/functional in this pull request? (as the new conditional tests are done in tests/new_functional)

**Review by ajmaddaford — CHANGES_REQUESTED** [review]

File: app/jinja_filters.py
Comment (berianjones) [reply app/jinja_filters.py]: Existence of date[0] checked, if no date[0], date is set to value of date[1]. Date is then checked and if None the exception is thrown. Thus the existence check on date[1] is carried out implicitly. A second check could be added, but this would add little value.

File: package.json
Comment (berianjones) [reply package.json]: OK

File: tests/new_functional/spec/dates_conditional.spec.js
Comment (berianjones) [reply tests/new_functional/spec/dates_conditional.spec.js]: done

File: tests/new_functional/pages/surveys/dates_conditional/summary.page.js
Comment (berianjones) [reply tests/new_functional/pages/surveys/dates_conditional/summary.page.js]: done

File: tests/functional/spec/dates.spec.js
Comment (berianjones) [reply tests/functional/spec/dates.spec.js]: contents are now identical with master branch version

File: app/jinja_filters.py
Comment (jonnyshaw89) [reply app/jinja_filters.py]: @berianjones line 58 would fail with `IndexError: list index out of range` so would not get to line 60 to be checked.

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: data/en/test_conditional_dates.json
Comment (berianjones) [reply data/en/test_conditional_dates.json:1]: done

File: app/jinja_filters.py
Comment (jonnyshaw89) [inline app/jinja_filters.py]: i think that if we make this method take 2 parameters then that would handle the None default for parameter 2.
you would just need to declare an additional `@blueprint.app_context_processor` for this

**Comment by berianjones** [comment]
Updated ninja_filters and test_jinja_filters to cater for error conditions of empty values passed, numbers rather than dates and date strings which are too short

**Comment by berianjones** [comment]
ready for review

File: app/jinja_filters.py
Comment (berianjones) [reply app/jinja_filters.py]: changes made to cater for this

File: app/jinja_filters.py
Comment (berianjones) [reply app/jinja_filters.py]: List index out of range error case addressed. Looking at 2 param option

File: tests/functional/pages/surveys/dates/dates-answers.page.js
Comment (berianjones) [reply tests/functional/pages/surveys/dates/dates-answers.page.js]: all removed

File: app/jinja_filters.py
Comment (ajmaddaford) [inline app/jinja_filters.py]: I don't think this Exception can ever get raised. Both `date1` and `date2` default to `None` if not passed, so dates will always be of length 2?

File: app/jinja_filters.py
Comment (ajmaddaford) [inline app/jinja_filters.py]: Why are we checking if either of the dates is an `int`?

File: app/jinja_filters.py
Comment (ajmaddaford) [inline app/jinja_filters.py]: This check is only relevant for strings, but placing it here will also check Date objects. Moving it to the else below would be better.

**Review by ajmaddaford — CHANGES_REQUESTED** [review]

File: app/jinja_filters.py
Comment (jonnyshaw89) [inline app/jinja_filters.py]: this code is pointless

File: app/jinja_filters.py
Comment (jonnyshaw89) [inline app/jinja_filters.py]: Might be worth updating this name to make it more obvious while we are making changes here.

File: tests/app/test_jinja_filters.py
Comment (jonnyshaw89) [inline tests/app/test_jinja_filters.py]: I think this indentation is incorrect so its only currently testing the last value in the list?

File: tests/app/test_jinja_filters.py
Comment (jonnyshaw89) [inline tests/app/test_jinja_filters.py]: same indentation issue?

**Review by jonnyshaw89 — CHANGES_REQUESTED** [review]

File: app/jinja_filters.py
Comment (berianjones) [reply app/jinja_filters.py]: Name updated, so that it reflects what its doing. I've changed the method format_condtional_date to accept two dates. Tests revised in consultation with Andrew.

File: tests/app/test_jinja_filters.py
Comment (berianjones) [reply tests/app/test_jinja_filters.py]: dealt with as part of test revision

File: tests/app/test_jinja_filters.py
Comment (berianjones) [reply tests/app/test_jinja_filters.py]: dealt with as part of test revision

**Comment by berianjones** [comment]
All tests passed, ready for review

File: app/jinja_filters.py
Comment (berianjones) [reply app/jinja_filters.py]: done

File: tests/app/test_jinja_filters.py
Comment (berianjones) [reply tests/app/test_jinja_filters.py]: changed

File: app/jinja_filters.py
Comment (berianjones) [reply app/jinja_filters.py]: removed as agreed

File: app/jinja_filters.py
Comment (berianjones) [reply app/jinja_filters.py]: removed as agreed

File: tests/app/test_jinja_filters.py
Comment (berianjones) [reply tests/app/test_jinja_filters.py]: changed

File: app/jinja_filters.py
Comment (berianjones) [reply app/jinja_filters.py]: removed now as part of param change

File: tests/functional/spec/dates.spec.js
Comment (berianjones) [reply tests/functional/spec/dates.spec.js]: added

File: data/en/test_dates.json
Comment (berianjones) [reply data/en/test_dates.json]: done

File: tests/app/test_jinja_filters.py
Comment (berianjones) [reply tests/app/test_jinja_filters.py]: changed now as calling using 2 date parameters not list

File: tests/app/test_jinja_filters.py
Comment (berianjones) [reply tests/app/test_jinja_filters.py]: added

File: tests/app/test_jinja_filters.py
Comment (berianjones) [reply tests/app/test_jinja_filters.py]: was added, but params now different

File: yarn.lock
Comment (berianjones) [reply yarn.lock]: reverted

**Review by ajmaddaford — APPROVED** [review]

**Comment by berianjones** [comment]
coverage issue addressed

**Comment by jonnyshaw89** [comment]
@berianjones you just need to revert your changes to `package.json` and `yarn.lock` then this is ready to go.

**Comment by jonnyshaw89** [comment]
@berianjones i don't think its the same as master. your master version may have been out of date.
Its still showing changes in this PR. possibly reverting some of the changes @adoublebarrel has recently made.

**Comment by berianjones** [comment]
@jonnyshaw89 can you review please. I've reverted package.json to @adoublebarrel 's changes

**Review by jonnyshaw89 — APPROVED** [review]

**Comment by jonnyshaw89** [comment]
@berianjones Can you squash your commits down to 1 before we merge?

**Pull request merged**
